# Segmenting and Clustering Neighborhoods in Toronto Question 1

### Web Scraping with BeautifulSoup

In [2]:
#imports
import pandas as pd
from bs4 import BeautifulSoup
import requests

In [3]:
#Obtaining HTML from website
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(source, 'lxml')

#Getting table info from HTML file
table = soup.find('table', {'class':'wikitable sortable'}).tbody

#Find all rows from table
rows = table.find_all('tr')

#Get column names
columns = [v.text.replace('\n', '') for v in rows[0].find_all('th')]

#initialize Dataframe with column names
df = pd.DataFrame(columns=columns)

#Populate Dataframe
for i in range(1, len(rows)):
    tds = rows[i].find_all('td')
    values = [tds[0].text.replace('\n',''), tds[1].text.replace('\n',''),tds[2].text.replace('\n','').replace(' /',',')]
    
    df = df.append(pd.Series(values, index=columns), ignore_index=True)
    
#Drop rows with Not assigned Borough
df = df[df.Borough != 'Not assigned']
#Reset index
df.reset_index(inplace=True,drop=True)
df.head(10)

Postal code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke                             Islington Avenue
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

In [4]:
df.shape

(103, 3)

# ***The shape of the data frame is (103, 3)***

## Assumptions/ Concerns
 - The ordering of the data frame was unclear, so it is ordered by Postal code
 - There were no cells that had a borough and a not assigned neighborhood, so no code was used to modify the data frame
     - Code below shows there are no such cells

In [5]:
test1 = df[df.Neighborhood == 'Not assigned']
test1.size

0

In [6]:
test2 = df[df.Neighborhood == '']
test2.size

0